# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1771243643 -39600
# Parent  b37b43c74945675c99ab7250e11134d39e387a29
NNN32 - Make sure 'After junk' filters also run on read IMAP messages.

diff --git a/mailnews/base/src/nsMsgDBFolder.cpp b/mailnews/base/src/nsMsgDBFolder.cpp
--- a/mailnews/base/src/nsMsgDBFolder.cpp
+++ b/mailnews/base/src/nsMsgDBFolder.cpp
@@ -2343,16 +2343,23 @@ nsMsgDBFolder::CallFilterPlugins(nsIMsgW
 
   MOZ_LOG(FILTERLOGMODULE, LogLevel::Info,
           ("Running filters on %" PRIu32 " new messages",
            (uint32_t)newKeys.Length()));
+  MOZ_LOG(FILTERLOGMODULE, LogLevel::Info,
+          ("Running filters on %" PRIu32 " saved new messages",
+           (uint32_t)m_saveNewMsgs.Length()));
 
   nsTArray<nsMsgKey> newMessageKeys;
   // Start from m_saveNewMsgs (and clear its current state).  m_saveNewMsgs is
   // where we stash the list of new messages when we are told to clear the list
   // of new messages by the UI (which purges the list from the nsMsgDatabase).
   newMessageKeys.SwapElements(m_saveNewMsgs);
   newMessageKeys.AppendElements(newKeys);
 
+  MOZ_LOG(FILTERLOGMODULE, LogLevel::Info,
+          ("Running filters on a total of %" PRIu32 " messages",
+           (uint32_t)newMessageKeys.Length()));
+
   // build up list of keys to classify
   nsTArray<nsMsgKey> classifyMsgKeys;
   nsCString uri;
 
diff --git a/mailnews/imap/src/nsImapMailFolder.cpp b/mailnews/imap/src/nsImapMailFolder.cpp
--- a/mailnews/imap/src/nsImapMailFolder.cpp
+++ b/mailnews/imap/src/nsImapMailFolder.cpp
@@ -3046,8 +3046,20 @@ nsresult nsImapMailFolder::NormalEndHead
                    "nsImapMailFolder::NormalEndHeaderParseStream()"));
           m_filterList->ApplyFiltersToHdr(
               nsMsgFilterType::InboxRule, newMsgHdr, this, mDatabase,
               nsDependentCSubstring(headers, headersSize), this, msgWindow);
+          // So here is a gigantic hack.
+          // Over in nsMsgDBFolder::CallFilterPlugins() we run 
+          // "after junk classification" filters on a "new list",
+          // but over there only unread messages are considered new.
+          // So if we need to run the filter on read messages, we need to
+          // communicate this somehow.
+          // Currently the message key is its UID, see above where it is set.
+          // Note that this will change:
+          // https://bugzilla.mozilla.org/show_bug.cgi?id=1806770
+          if (filterOnHighwater && (msgFlags & nsMsgMessageFlags::Read)) {
+            mDatabase->AddToNewList(m_curMsgUid);
+          }
           NotifyFolderEvent(kFiltersApplied);
         }
       }
     }
